### PR TITLE
Forward flow direction to the teaching tip content root

### LINKED ISSUE #458: TeachingTip in RTL treats Placement=Left as physical instead of logical
Forking off this issue from #390 since this is one thing that we know we need to handle. The rest of #390 needs a bit more information/investigation before it's fully actionable.

### PR DESCRIPTION
**Pull request by StephenLPeters**
Forward flow direction to the teaching tip content root and adjust placement when flow direction is set rightToLeft. Fixes #458
Cast to IPopup3 instead of checking version. 
Make the automation peer public.

### CODE REVIEW

File: dev/TeachingTip/TeachingTip.cpp [hunk 1]
```diff
@@ -439,23 +439,24 @@ bool TeachingTip::PositionTargetedPopup()
     double tipWidth = tailOcclusionGrid.ActualWidth();
 
     auto&& popup = m_popup.get();
+    auto isLeftToRight = FlowDirection() == winrt::FlowDirection::LeftToRight;
     // Depending on the effective placement mode of the tip we use a combination of the tip's size, the target's position within the app, the target's
```
Comment (RBrid) [inline dev/TeachingTip/TeachingTip.cpp]: How about doing this once only?
```
if (!isLeftToRight)
{
    tipWidth *= -1.0;
}

```

File: dev/TeachingTip/TeachingTip.cpp [hunk 2]
```diff
@@ -923,8 +926,8 @@ void TeachingTip::OnShouldConstrainToRootBoundsChanged()
     // If we have opened the tip's popup and then this property changes we will need to discard the old popup
     // and replace it with a new popup.  This variable indicates this state.
 
-    //The underlying popup api is only available on 19h1 plus, if we aren't on that no opt.
-    if (SharedHelpers::Is19H1OrHigher())
+    // If the popup isn't created or doesn't have the ShouldConstrainToRootBounds api then no opt
+    if(m_popup.get().try_as<winrt::Controls::Primitives::IPopup3>())
```
Comment (chrisglein) [inline dev/TeachingTip/TeachingTip.cpp]:
```suggestion
    if (m_popup.get().try_as<winrt::Controls::Primitives::IPopup3>())
```

File: dev/TeachingTip/TestUI/TeachingTipPage.xaml [hunk 3]
```diff
@@ -30,7 +30,8 @@
                                           Closing="OnTeachingTipClosing"
                                           ActionButtonClick="OnTeachingTipActionButtonClicked"
                                           CloseButtonClick="OnTeachingTipCloseButtonClicked"
-                                          Target="{x:Bind targetButton}">
+                                          Target="{x:Bind targetButton}"
+                                          FlowDirection="RightToLeft">
```
Comment (chrisglein) [inline dev/TeachingTip/TestUI/TeachingTipPage.xaml]: Did you mean to have this left as RTL in the test app?

File: dev/TeachingTip/TeachingTip.cpp [hunk 4]
```diff
@@ -529,72 +530,73 @@ bool TeachingTip::PositionUntargetedPopup()
     // Depending on the effective placement mode of the tip we use a combination of the tip's size, the window's size, and the target
     // offset property to determine the appropriate vertical and horizontal offsets of the popup that the tip is contained in.
     auto&& popup = m_popup.get();
+    auto isLeftToRight = FlowDirection() == winrt::FlowDirection::LeftToRight;
     switch (PreferredPlacement())
```
Comment (RBrid) [inline dev/TeachingTip/TeachingTip.cpp]: Same comment here. If all usages of finalTipWidth are affected by isLeftToRight, you can just reverse its sign once.

File: dev/TeachingTip/TestUI/TeachingTipPage.xaml (hunk 3 above)
Comment (StephenLPeters) [reply dev/TeachingTip/TestUI/TeachingTipPage.xaml]: I did, we have two tips, one in RTL one in LTR

---
In reply to: [270511702](https://github.com/Microsoft/microsoft-ui-xaml/pull/504#discussion_r270511702) [](ancestors = 270511702)

File: dev/TeachingTip/TeachingTip.cpp (hunk 2 above)
Comment (RBrid) [inline dev/TeachingTip/TeachingTip.cpp]: nit: missing space.

Comment (StephenLPeters) [reply dev/TeachingTip/TeachingTip.cpp]: Isn't this what is already there?

---
In reply to: [270511566](https://github.com/Microsoft/microsoft-ui-xaml/pull/504#discussion_r270511566) [](ancestors = 270511566)

File: dev/TeachingTip/TeachingTip.cpp (hunk 4 above)
Comment (StephenLPeters) [reply dev/TeachingTip/TeachingTip.cpp]: This one is less enticing, for Near and Far placements we want 0 in one case and Positive in the other, while in center placement we want negative.

---
In reply to: [270512149](https://github.com/Microsoft/microsoft-ui-xaml/pull/504#discussion_r270512149) [](ancestors = 270512149)

**Pull request closed without merging**